libvast/vast/logger.hpp

```diff
@@ -45,33 +45,40 @@ void fixup_logger(const system::configuration& cfg);
 
 #define VAST_LOG_2(lvl, m1) VAST_LOG_IMPL(lvl, m1)
 
-#define VAST_LOG_3(lvl, m1, m2) VAST_LOG_IMPL(lvl, m1 << m2)
+#  define VAST_LOG_3(lvl, m1, m2) VAST_LOG_IMPL(lvl, (m1) << (m2))
 
-#define VAST_LOG_4(lvl, m1, m2, m3) VAST_LOG_IMPL(lvl, m1 << m2 << m3)
+#  define VAST_LOG_4(lvl, m1, m2, m3) VAST_LOG_IMPL(lvl, (m1) << (m2) << (m3))
 
-#define VAST_LOG_5(lvl, m1, m2, m3, m4) VAST_LOG_IMPL(lvl, m1 << m2 << m3 << m4)
+#  define VAST_LOG_5(lvl, m1, m2, m3, m4)                                      \
+    VAST_LOG_IMPL(lvl, (m1) << (m2) << (m3) << (m4))
 
-#define VAST_LOG_6(lvl, m1, m2, m3, m4, m5)                                    \
-  VAST_LOG_IMPL(lvl, m1 << m2 << m3 << m4 << m5)
+#  define VAST_LOG_6(lvl, m1, m2, m3, m4, m5)                                  \
+    VAST_LOG_IMPL(lvl, (m1) << (m2) << (m3) << (m4) << (m5))
 
-#define VAST_LOG_7(lvl, m1, m2, m3, m4, m5, m6)                                \
-  VAST_LOG_IMPL(lvl, m1 << m2 << m3 << m4 << m5 << m6)
+#  define VAST_LOG_7(lvl, m1, m2, m3, m4, m5, m6)                              \
+    VAST_LOG_IMPL(lvl, (m1) << (m2) << (m3) << (m4) << (m5) << (m6))
 
-#define VAST_LOG_8(lvl, m1, m2, m3, m4, m5, m6, m7)                            \
-  VAST_LOG_IMPL(lvl, m1 << m2 << m3 << m4 << m5 << m6 << m7)
+#  define VAST_LOG_8(lvl, m1, m2, m3, m4, m5, m6, m7)                          \
+    VAST_LOG_IMPL(lvl, (m1) << (m2) << (m3) << (m4) << (m5) << (m6) << (m7))
 
-#define VAST_LOG_9(lvl, m1, m2, m3, m4, m5, m6, m7, m8)                        \
-  VAST_LOG_IMPL(lvl, m1 << m2 << m3 << m4 << m5 << m6 << m7 << m8)
+#  define VAST_LOG_9(lvl, m1, m2, m3, m4, m5, m6, m7, m8)                      \
+    VAST_LOG_IMPL(lvl, (m1) << (m2) << (m3) << (m4) << (m5) << (m6) << (m7)    \
+                            << (m8))
 
-#define VAST_LOG_10(lvl, m1, m2, m3, m4, m5, m6, m7, m8, m9)                   \
-  VAST_LOG_IMPL(lvl, m1 << m2 << m3 << m4 << m5 << m6 << m7 << m8 << m9)
+#  define VAST_LOG_10(lvl, m1, m2, m3, m4, m5, m6, m7, m8, m9)                 \
+    VAST_LOG_IMPL(lvl, (m1) << (m2) << (m3) << (m4) << (m5) << (m6) << (m7)    \
+                            << (m8) << (m9))
 
 #  define VAST_LOG_11(lvl, m1, m2, m3, m4, m5, m6, m7, m8, m9, m10)            \
-    VAST_LOG_IMPL(lvl,                                                         \
-                  m1 << m2 << m3 << m4 << m5 << m6 << m7 << m8 << m9 << m10)
+    VAST_LOG_IMPL(lvl, (m1) << (m2) << (m3) << (m4) << (m5) << (m6) << (m7)    \
+                            << (m8) << (m9) << (m10))
 
 #  define VAST_LOG(...) VAST_PP_OVERLOAD(VAST_LOG_, __VA_ARGS__)(__VA_ARGS__)
 
+// Ensures all args are used and syntactically valid, without evaluating them.
+#  define VAST_LOG_DISCARD_ARGS(...)                                           \
+    { [[maybe_unused]] decltype(__VA_ARGS__, nullptr) unused{}; }
```

Comment:
The current macro has issues with code like this:

```cpp
if (...) 
  VAST_WARNING(...)
else
  ...
```

Because it gets expanded to this:

```cpp
if (...)
  { [[maybe_unused]] decltype(..., nulltr) unused{}; };
else
  ...
```

The semicolon after the braces is causing issues here.

You can fix this by using this instead:

```suggestion
#  define VAST_LOG_DISCARD_ARGS(...)                                           \
    do { [[maybe_unused]] decltype(__VA_ARGS__, 0) unused{}; } while(false)
```

(Only changed from `nullptr` to `0` so it still fits the same line. Really doesn't matter what we have there.)